src/mod/dns.mod/coredns.c

```diff
@@ -1201,6 +1204,103 @@ static void dns_lookup(sockname_t *addr)
   sendrequest(rp, T_PTR);
 }
 
+static int dns_hosts(char *hostn) {
+  #define PATH "/etc/hosts"
+  int fd, len, i, found = 0;
+  struct stat sb;
+  char *addr, hostn_lower[256], hostn_upper[256], *last_newline, *last_hash = NULL, *c, *c2, ip[256];
+  sockname_t name;
+
+  if (!*hostn) {
+    ddebug0(RES_MSG "bogus empty hostname input");
+    return 1;
+  }
+  len = strlen(hostn);
+  if (len > 255) {
+    ddebug0(RES_MSG "bogus len of hostname > 255 input");
+    return 1;
+  }
+  /* due to strncasecmp() and strncmp() being slow if used in loop, precalculate
+   * lower and upper string from hostn and compare with handcrafted code */
+  for (i = 0; i < len; i++) {
+      if (isspace(hostn[i])) {
+        ddebug0(RES_MSG "bogus white-space hostname input");
+        return 1;
+      }
+      hostn_lower[i] = tolower(hostn[i]);
+      hostn_upper[i] = toupper(hostn[i]);
+  }
+  hostn_lower[i] = 0;
+  hostn_upper[i] = 0;
+  fd = open(PATH, O_RDONLY);
+  if (fd < 0) {
+    ddebug1(RES_MSG "open(" PATH "): %s ", strerror(errno));
+    return 0;
+  }
+  if (fstat(fd, &sb) < 0) {
+    ddebug1(RES_MSG "fstat(" PATH "): %s ", strerror(errno));
+    close(fd);
+    return 0;
+  }
+  addr = mmap(NULL, sb.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+  if (addr == MAP_FAILED) {
+    ddebug1(RES_MSG "mmap(" PATH "): %s ", strerror(errno));
+    close(fd);
+    return 0;
+  }
+  last_newline = addr - 1;
+  /* case insensitive search for hostn */
+  for (c = addr; (c < (addr + sb.st_size - len)) && *c; c++) {
+    switch(*c) {
+      case '\n':
+      case '\r':
+        last_newline = c;
+        break;
+      case '#':
+        last_hash = c;
+        break;
+      default:
+        /* if (!strncasecmp(c, hostn, len)) { */
+        for (i = 0; (i < len) && (c[i] == hostn_lower[i] || (c[i] == hostn_upper[i])); i++);
+        if ((i == len) &&
+            ((c == (addr + sb.st_size - len - 1)) || isspace(*(c + len))) &&
```

Comment:
this can again be invalid (it ca be addr - 1). Example:
file "ex" and we are looking for "ex", addr points to "e", sb.st_size is 2, len is 2: addr + 2 - 2 - 1 => addr - 1 which is undefined behavior to calculate.
Is it correct? I think you intend to check if it's the end of the file, addr - 1 would end up wrong.
isspace needs the unsigned char cast as well

Reply:
as we "agreed" we read this file (which can be bigger than 1k ;) for every dns lookup, not caching (looking at the files modification time). so i thought it would be worth using mmap. but yes, i will change it to using line by line reading now. esp. since the future will hopefully bring async dns with #1015 and make this dns mod code obsolete.

Reply:
Fixed isspace() by using the already existing egg_isspace() macro